test/integration/apiserver/flowcontrol/concurrency_test.go

```diff
@@ -0,0 +1,249 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/master"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	dispatchedRequestCountMetricsName               = "apiserver_flowcontrol_dispatched_requests_total"
+	dispatchedRequestCountMetricsLabelPriorityLevel = "priorityLevel"
+	timeout                                         = time.Second * 10
+)
+
+func setup(t testing.TB) (*httptest.Server, *rest.Config, framework.CloseFunc) {
+	opts := framework.MasterConfigOptions{EtcdOptions: framework.DefaultEtcdOptions()}
+	opts.EtcdOptions.DefaultStorageMediaType = "application/vnd.kubernetes.protobuf"
+	masterConfig := framework.NewIntegrationTestMasterConfigWithOptions(&opts)
+	resourceConfig := master.DefaultAPIResourceConfigSource()
+	resourceConfig.EnableVersions(schema.GroupVersion{
+		Group:   "flowcontrol.apiserver.k8s.io",
+		Version: "v1alpha1",
+	})
+	masterConfig.GenericConfig.MaxRequestsInFlight = 5
+	masterConfig.GenericConfig.MaxMutatingRequestsInFlight = 5
+	masterConfig.GenericConfig.OpenAPIConfig = framework.DefaultOpenAPIConfig()
+	masterConfig.ExtraConfig.APIResourceConfigSource = resourceConfig
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+
+	return s, masterConfig.GenericConfig.LoopbackClientConfig, closeFn
+}
+
+func TestPriorityLevelIsolation(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.APIPriorityAndFairness, true)()
+	// NOTE: disabling the feature should fail the test
+	_, loopbackConfig, closeFn := setup(t)
+	defer closeFn()
+
+	loopbackClient := clientset.NewForConfigOrDie(loopbackConfig)
+	noxu1Client := getClientFor(loopbackConfig, "noxu1")
+	noxu2Client := getClientFor(loopbackConfig, "noxu2")
+
+	priorityLevelNoxu1, _, err := createPriorityLevelAndBindingFlowSchemaForUser(loopbackClient, "noxu1")
+	require.NoError(t, err)
+	priorityLevelNoxu2, _, err := createPriorityLevelAndBindingFlowSchemaForUser(loopbackClient, "noxu2")
+	require.NoError(t, err)
+
+	wg := &sync.WaitGroup{}
+	// "elephant"
+	streamRequests(wg, 10, 100, func() {
+		_, err := noxu1Client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+		require.NoError(t, err)
+	})
+
+	streamRequests(nil, 1, 100, func() {
+		_, err := noxu2Client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+		require.NoError(t, err)
+	})
+
+	wg.Wait()
+
+	dispatchedCountNoxu1, err := getRequestCountOfPriorityLevel(loopbackClient, priorityLevelNoxu1.Name)
+	require.NoError(t, err)
+	dispatchedCountNoxu2, err := getRequestCountOfPriorityLevel(loopbackClient, priorityLevelNoxu2.Name)
+	require.NoError(t, err)
+
+	assert.Equal(t, 1000, dispatchedCountNoxu1)
+	assert.Equal(t, 100, dispatchedCountNoxu2)
```

Comment:
I don't understand what this is actually testing, it seems like we need to say something about the time taken and not just the requests sent.

I think we should do two tests, one where neither client is limited, and one where they are both limited, and show that in the latter case they both get the same number of requests?

Reply:
the point is to verify that their requests are executing in separated bands, b/c the two clients are matching two different priority-level. in this test, by putting an elephant (1k) into one priority and a mouse (100) into another, we are supposed to at least observe that the mouse completes before the elephants. on the other hand, w/ the feature disabled, all the requests from both clients will be executing in a shared band w/o fairness, the elephant will be eating up all the bandwidth.

>  it seems like we need to say something about the time taken and not just the requests sent.

didn't we agree on not using service-time to avoid scheduling/performance noises?

> I think we should do two tests, one where neither client is limited, and one where they are both limited, and show that in the latter case they both get the same number of requests?

that's a good idea for testing fairness within one priority-level, but different points from which the current test is making.

Reply:
The test here seems to be intended to test that engaging the feature converts a situation where the max-in-flight limit of 5 --- which would produce 429 replies due to the 11 clients and trigger the error detection in the test func --- is converted into two concurrency limits of 1 (I am assuming the suggested configuration gets injected in this test) plus enough queuing to cause no rejections.

Since each priority level has a queue length limit of 10 and a concurrency limit of (at least) 1, that means it has a capacity for 11 concurrent clients.  And 11 is the total nunber of concurrent clients in this test.  So this test does not really test that the two categories of clients got separated.

Reply:
> (I am assuming the suggested configuration gets injected in this test)

we double-checks the flow-schema get injected by checking its status after creation

> Since each priority level has a queue length limit of 10 and a concurrency limit of (at least) 1

yes, each testing priority-level actually gets 1 concurrency.

> Since each priority level has a queue length limit of 10 and a concurrency limit of (at least) 1, that means it has a capacity for 11 concurrent clients.

i updated the concurrency of elephants to 20, so it's 21 total requesting concurrency > 11. wdyt?

Reply:
Ah, on my first read, I missed that the waitgroup wasn't waiting for both clients. So, the time permitted for the mouse is limited by the time taken for the elephant. That's what I was looking for.

I agree making it actually exceed the limits seems necessary (in fact, could consider testing it both ways).

Reply:
@yue9944882:

- I was referring to https://github.com/kubernetes/kubernetes/blob/d00f9c7c1091e31c75c6636500095c4e490b8db8/staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go#L46-L74 .  Absent them, the only other concurrency shares will be the 1 for the catch-all --- in which case the elephant and mouse will each get a concurrency limit of 5.

- 21 clients now exceeds the single-class capacity of 11.  In fact, 20 clients exceeds 11 --- on which I also remarked inline.

@lavalamp: neither mouse nor elephant is time-limited by the other.  The WaitGroup is used to wait until they are both done, not until the first is done.